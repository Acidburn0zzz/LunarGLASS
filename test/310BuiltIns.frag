#version 310 es

precision mediump float;

uniform int imax, imin;
uniform uint umax, umin;

in vec3 x, y;
uniform bvec3 bv;

flat in uint uy;
flat in uvec2 uv2c;
flat in uvec2 uv2y;
flat in uvec2 uv2x;
flat in uvec4 uv4y;

flat in ivec3 iv3a;
flat in ivec3 iv3b;

flat in ivec4 iv4a;
flat in ivec4 iv4b;

in float f;

in vec2 v2a, v2b;

in vec4 v4;

out vec4 result;

void main()
{
    uvec4 usum = uvec4(0);
    ivec4 isum = ivec4(0);
    vec4 sum = vec4(0);

    // 8.3 int
    sum += vec4(mix(iv3a, iv3b, bv), 0.0);

    result = sum + vec4(isum) + vec4(usum);

    result += gl_FragCoord + vec4(gl_PointCoord, 2.0 * gl_PointCoord);
    gl_FragDepth = f;
    gl_FragDepth += 0.1;
}
