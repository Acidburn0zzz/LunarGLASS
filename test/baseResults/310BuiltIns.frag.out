
Top IR:
; ModuleID = 'Glslang'

@iv3a = global <3 x i32> zeroinitializer
@iv3b = global <3 x i32> zeroinitializer
@bv = external addrspace(2) constant <3 x i1>
@result = global <4 x float> zeroinitializer
@gl_FragCoord = global <4 x float> zeroinitializer
@gl_PointCoord = global <2 x float> zeroinitializer
@gl_FragDepth = global float 0.000000e+00
@f = global float 0.000000e+00
@imax = external addrspace(2) constant i32
@imin = external addrspace(2) constant i32
@umax = external addrspace(2) constant i32
@umin = external addrspace(2) constant i32
@x = global <3 x float> zeroinitializer
@y = global <3 x float> zeroinitializer
@uy = global i32 0
@uv2c = global <2 x i32> zeroinitializer
@uv2y = global <2 x i32> zeroinitializer
@uv2x = global <2 x i32> zeroinitializer
@uv4y = global <4 x i32> zeroinitializer
@iv4a = global <4 x i32> zeroinitializer
@iv4b = global <4 x i32> zeroinitializer
@v2a = global <2 x float> zeroinitializer
@v2b = global <2 x float> zeroinitializer
@v4 = global <4 x float> zeroinitializer

define fastcc void @main() {
entry:
  %result3 = alloca <4 x float>
  %sum1 = alloca <4 x float>
  %sum = alloca <4 x float>
  %isum = alloca <4 x i32>
  %usum = alloca <4 x i32>
  br label %mainBody

mainBody:                                         ; preds = %entry
  store <4 x i32> zeroinitializer, <4 x i32>* %usum
  store <4 x i32> zeroinitializer, <4 x i32>* %isum
  store <4 x float> zeroinitializer, <4 x float>* %sum
  %0 = load <3 x i32>* @iv3a
  %1 = load <3 x i32>* @iv3b
  %2 = load <3 x i1> addrspace(2)* @bv, !gla.uniform !35
  %3 = select <3 x i1> %2, <3 x i32> %1, <3 x i32> %0
  %4 = sitofp <3 x i32> %3 to <3 x float>, !gla.precision !47
  %5 = load <4 x float>* %sum1
  %6 = extractelement <3 x float> %4, i32 0, !gla.precision !47
  %7 = insertelement <4 x float> %5, float %6, i32 0, !gla.precision !47
  %8 = extractelement <3 x float> %4, i32 1, !gla.precision !47
  %9 = insertelement <4 x float> %7, float %8, i32 1, !gla.precision !47
  %10 = extractelement <3 x float> %4, i32 2, !gla.precision !47
  %11 = insertelement <4 x float> %9, float %10, i32 2, !gla.precision !47
  %12 = insertelement <4 x float> %11, float 0.000000e+00, i32 3, !gla.precision !47
  %13 = load <4 x float>* %sum
  %sum2 = fadd <4 x float> %13, %12, !gla.precision !47
  store <4 x float> %sum2, <4 x float>* %sum
  %14 = load <4 x float>* %sum
  %15 = load <4 x i32>* %isum
  %16 = sitofp <4 x i32> %15 to <4 x float>, !gla.precision !47
  %17 = fadd <4 x float> %14, %16, !gla.precision !47
  %18 = load <4 x i32>* %usum
  %19 = uitofp <4 x i32> %18 to <4 x float>, !gla.precision !47
  %result = fadd <4 x float> %17, %19, !gla.precision !47
  store <4 x float> %result, <4 x float>* @result
  %20 = load <4 x float>* @gl_FragCoord
  %21 = load <2 x float>* @gl_PointCoord
  %22 = load <2 x float>* @gl_PointCoord
  %23 = fmul <2 x float> <float 2.000000e+00, float 2.000000e+00>, %22, !gla.precision !47
  %24 = load <4 x float>* %result3
  %25 = extractelement <2 x float> %21, i32 0, !gla.precision !48
  %26 = insertelement <4 x float> %24, float %25, i32 0, !gla.precision !48
  %27 = extractelement <2 x float> %21, i32 1, !gla.precision !48
  %28 = insertelement <4 x float> %26, float %27, i32 1, !gla.precision !48
  %29 = extractelement <2 x float> %23, i32 0, !gla.precision !48
  %30 = insertelement <4 x float> %28, float %29, i32 2, !gla.precision !48
  %31 = extractelement <2 x float> %23, i32 1, !gla.precision !48
  %32 = insertelement <4 x float> %30, float %31, i32 3, !gla.precision !48
  %33 = fadd <4 x float> %20, %32, !gla.precision !48
  %34 = load <4 x float>* @result
  %result4 = fadd <4 x float> %34, %33, !gla.precision !48
  store <4 x float> %result4, <4 x float>* @result
  %gl_FragDepth = load float* @f
  store float %gl_FragDepth, float* @gl_FragDepth
  %35 = load float* @gl_FragDepth
  %gl_FragDepth5 = fadd float %35, 0x3FB99999A0000000, !gla.precision !48
  store float %gl_FragDepth5, float* @gl_FragDepth
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %mainBody
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

!gla.entrypoint = !{!0}
!gla.inputs = !{!1, !3, !5, !7, !9, !11, !13, !15, !17, !19, !21, !23, !25, !27, !29, !31, !33}
!gla.uniforms = !{!35, !37, !39, !40, !42}
!gla.outputs = !{!43, !45}
!gla.noStaticUse = !{!37, !39, !40, !42, !11, !13, !15, !17, !19, !21, !23, !25, !27, !29, !31, !33}

!0 = !{!"main", i32 15}
!1 = !{!"iv3a", i32 1, <3 x i32>* @iv3a_typeProxy, !2, !""}
!2 = !{i32 0, i32 2, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!3 = !{!"iv3b", i32 1, <3 x i32>* @iv3b_typeProxy, !4, !""}
!4 = !{i32 0, i32 2, i32 1025, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!5 = !{!"gl_FragCoord", i32 5, <4 x float>* @gl_FragCoord_typeProxy, !6, !""}
!6 = !{i32 0, i32 3, i32 1027, null, i32 0, i32 41, i32 -1, i32 0, i32 -1}
!7 = !{!"gl_PointCoord", i32 6, <2 x float>* @gl_PointCoord_typeProxy, !8, !""}
!8 = !{i32 0, i32 2, i32 1028, null, i32 0, i32 42, i32 -1, i32 0, i32 -1}
!9 = !{!"f", i32 1, float* @f_typeProxy, !10, !""}
!10 = !{i32 0, i32 2, i32 1030, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!11 = !{!"x", i32 1, <3 x float>* @x_typeProxy, !12, !""}
!12 = !{i32 0, i32 2, i32 1031, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!13 = !{!"y", i32 1, <3 x float>* @y_typeProxy, !14, !""}
!14 = !{i32 0, i32 2, i32 1032, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!15 = !{!"uy", i32 1, i32* @uy_typeProxy, !16, !""}
!16 = !{i32 1, i32 2, i32 1033, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!17 = !{!"uv2c", i32 1, <2 x i32>* @uv2c_typeProxy, !18, !""}
!18 = !{i32 1, i32 2, i32 1034, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!19 = !{!"uv2y", i32 1, <2 x i32>* @uv2y_typeProxy, !20, !""}
!20 = !{i32 1, i32 2, i32 1035, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!21 = !{!"uv2x", i32 1, <2 x i32>* @uv2x_typeProxy, !22, !""}
!22 = !{i32 1, i32 2, i32 1036, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!23 = !{!"uv4y", i32 1, <4 x i32>* @uv4y_typeProxy, !24, !""}
!24 = !{i32 1, i32 2, i32 1037, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!25 = !{!"iv4a", i32 1, <4 x i32>* @iv4a_typeProxy, !26, !""}
!26 = !{i32 0, i32 2, i32 1038, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!27 = !{!"iv4b", i32 1, <4 x i32>* @iv4b_typeProxy, !28, !""}
!28 = !{i32 0, i32 2, i32 1039, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = !{!"v2a", i32 1, <2 x float>* @v2a_typeProxy, !30, !""}
!30 = !{i32 0, i32 2, i32 1040, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!31 = !{!"v2b", i32 1, <2 x float>* @v2b_typeProxy, !32, !""}
!32 = !{i32 0, i32 2, i32 1041, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!33 = !{!"v4", i32 1, <4 x float>* @v4_typeProxy, !34, !""}
!34 = !{i32 0, i32 2, i32 1042, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!35 = !{!"bv", i32 12, <3 x i1>* @bv_typeProxy, !36, !""}
!36 = !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!37 = !{!"imax", i32 12, i32* @imax_typeProxy, !38, !""}
!38 = !{i32 0, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!39 = !{!"imin", i32 12, i32* @imin_typeProxy, !38, !""}
!40 = !{!"umax", i32 12, i32* @umax_typeProxy, !41, !""}
!41 = !{i32 1, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!42 = !{!"umin", i32 12, i32* @umin_typeProxy, !41, !""}
!43 = !{!"result", i32 7, <4 x float>* @result_typeProxy, !44, !""}
!44 = !{i32 0, i32 2, i32 1026, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!45 = !{!"gl_FragDepth", i32 11, float* @gl_FragDepth_typeProxy, !46, !""}
!46 = !{i32 0, i32 3, i32 1029, null, i32 0, i32 45, i32 -1, i32 0, i32 -1}
!47 = !{i32 2}
!48 = !{i32 3}


Bottom IR:
; ModuleID = 'Glslang'
target datalayout = "e-p:32:32"

@iv3a = global <3 x i32> zeroinitializer
@iv3b = global <3 x i32> zeroinitializer
@bv = external addrspace(2) constant <3 x i1>
@result = global <4 x float> zeroinitializer
@gl_FragCoord = global <4 x float> zeroinitializer
@gl_PointCoord = global <2 x float> zeroinitializer
@gl_FragDepth = global float 0.000000e+00
@f = global float 0.000000e+00
@x = global <3 x float> zeroinitializer
@y = global <3 x float> zeroinitializer
@uy = global i32 0
@uv2c = global <2 x i32> zeroinitializer
@uv2y = global <2 x i32> zeroinitializer
@uv2x = global <2 x i32> zeroinitializer
@uv4y = global <4 x i32> zeroinitializer
@iv4a = global <4 x i32> zeroinitializer
@iv4b = global <4 x i32> zeroinitializer
@v2a = global <2 x float> zeroinitializer
@v2b = global <2 x float> zeroinitializer
@v4 = global <4 x float> zeroinitializer

define fastcc void @main() {
entry:
  %0 = load <3 x i32>* @iv3a
  %1 = load <3 x i32>* @iv3b
  %2 = load <3 x i1> addrspace(2)* @bv, !gla.uniform !35
  %3 = select <3 x i1> %2, <3 x i32> %1, <3 x i32> %0
  %4 = sitofp <3 x i32> %3 to <3 x float>, !gla.precision !47
  %5 = call <4 x float> @llvm.gla.fMultiInsert.v4f32.v4f32.v3f32.v3f32.v3f32.f32(<4 x float> undef, i32 15, <3 x float> %4, i32 0, <3 x float> %4, i32 1, <3 x float> %4, i32 2, float 0.000000e+00, i32 0)
  %sum2 = fadd <4 x float> %5, zeroinitializer, !gla.precision !47
  %6 = fadd <4 x float> %sum2, zeroinitializer, !gla.precision !47
  %result = fadd <4 x float> %6, zeroinitializer, !gla.precision !47
  store <4 x float> %result, <4 x float>* @result
  %7 = load <4 x float>* @gl_FragCoord
  %8 = load <2 x float>* @gl_PointCoord
  %9 = fmul <2 x float> %8, <float 2.000000e+00, float 2.000000e+00>, !gla.precision !47
  %10 = call <4 x float> @llvm.gla.fMultiInsert.v4f32.v4f32.v2f32.v2f32.v2f32.v2f32(<4 x float> undef, i32 15, <2 x float> %8, i32 0, <2 x float> %8, i32 1, <2 x float> %9, i32 0, <2 x float> %9, i32 1)
  %11 = fadd <4 x float> %7, %10, !gla.precision !48
  %result4 = fadd <4 x float> %11, %result, !gla.precision !48
  store <4 x float> %result4, <4 x float>* @result
  %gl_FragDepth = load float* @f
  %gl_FragDepth5 = fadd float %gl_FragDepth, 0x3FB99999A0000000, !gla.precision !48
  store float %gl_FragDepth5, float* @gl_FragDepth
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %entry
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: nounwind readnone
declare <4 x float> @llvm.gla.fMultiInsert.v4f32.v4f32.v2f32.v2f32.v2f32.v2f32(<4 x float>, i32, <2 x float>, i32, <2 x float>, i32, <2 x float>, i32, <2 x float>, i32) #0

; Function Attrs: nounwind readnone
declare <4 x float> @llvm.gla.fMultiInsert.v4f32.v4f32.v3f32.v3f32.v3f32.f32(<4 x float>, i32, <3 x float>, i32, <3 x float>, i32, <3 x float>, i32, float, i32) #0

attributes #0 = { nounwind readnone }

!gla.entrypoint = !{!0}
!gla.inputs = !{!1, !3, !5, !7, !9, !11, !13, !15, !17, !19, !21, !23, !25, !27, !29, !31, !33}
!gla.uniforms = !{!35, !37, !39, !40, !42}
!gla.outputs = !{!43, !45}
!gla.noStaticUse = !{!37, !39, !40, !42, !11, !13, !15, !17, !19, !21, !23, !25, !27, !29, !31, !33}

!0 = !{!"main", i32 15}
!1 = !{!"iv3a", i32 1, <3 x i32>* @iv3a_typeProxy, !2, !""}
!2 = !{i32 0, i32 2, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!3 = !{!"iv3b", i32 1, <3 x i32>* @iv3b_typeProxy, !4, !""}
!4 = !{i32 0, i32 2, i32 1025, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!5 = !{!"gl_FragCoord", i32 5, <4 x float>* @gl_FragCoord_typeProxy, !6, !""}
!6 = !{i32 0, i32 3, i32 1027, null, i32 0, i32 41, i32 -1, i32 0, i32 -1}
!7 = !{!"gl_PointCoord", i32 6, <2 x float>* @gl_PointCoord_typeProxy, !8, !""}
!8 = !{i32 0, i32 2, i32 1028, null, i32 0, i32 42, i32 -1, i32 0, i32 -1}
!9 = !{!"f", i32 1, float* @f_typeProxy, !10, !""}
!10 = !{i32 0, i32 2, i32 1030, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!11 = !{!"x", i32 1, <3 x float>* @x_typeProxy, !12, !""}
!12 = !{i32 0, i32 2, i32 1031, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!13 = !{!"y", i32 1, <3 x float>* @y_typeProxy, !14, !""}
!14 = !{i32 0, i32 2, i32 1032, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!15 = !{!"uy", i32 1, i32* @uy_typeProxy, !16, !""}
!16 = !{i32 1, i32 2, i32 1033, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!17 = !{!"uv2c", i32 1, <2 x i32>* @uv2c_typeProxy, !18, !""}
!18 = !{i32 1, i32 2, i32 1034, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!19 = !{!"uv2y", i32 1, <2 x i32>* @uv2y_typeProxy, !20, !""}
!20 = !{i32 1, i32 2, i32 1035, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!21 = !{!"uv2x", i32 1, <2 x i32>* @uv2x_typeProxy, !22, !""}
!22 = !{i32 1, i32 2, i32 1036, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!23 = !{!"uv4y", i32 1, <4 x i32>* @uv4y_typeProxy, !24, !""}
!24 = !{i32 1, i32 2, i32 1037, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!25 = !{!"iv4a", i32 1, <4 x i32>* @iv4a_typeProxy, !26, !""}
!26 = !{i32 0, i32 2, i32 1038, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!27 = !{!"iv4b", i32 1, <4 x i32>* @iv4b_typeProxy, !28, !""}
!28 = !{i32 0, i32 2, i32 1039, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = !{!"v2a", i32 1, <2 x float>* @v2a_typeProxy, !30, !""}
!30 = !{i32 0, i32 2, i32 1040, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!31 = !{!"v2b", i32 1, <2 x float>* @v2b_typeProxy, !32, !""}
!32 = !{i32 0, i32 2, i32 1041, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!33 = !{!"v4", i32 1, <4 x float>* @v4_typeProxy, !34, !""}
!34 = !{i32 0, i32 2, i32 1042, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!35 = !{!"bv", i32 12, <3 x i1>* @bv_typeProxy, !36, !""}
!36 = !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!37 = !{!"imax", i32 12, i32* @imax_typeProxy, !38, !""}
!38 = !{i32 0, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!39 = !{!"imin", i32 12, i32* @imin_typeProxy, !38, !""}
!40 = !{!"umax", i32 12, i32* @umax_typeProxy, !41, !""}
!41 = !{i32 1, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!42 = !{!"umin", i32 12, i32* @umin_typeProxy, !41, !""}
!43 = !{!"result", i32 7, <4 x float>* @result_typeProxy, !44, !""}
!44 = !{i32 0, i32 2, i32 1026, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!45 = !{!"gl_FragDepth", i32 11, float* @gl_FragDepth_typeProxy, !46, !""}
!46 = !{i32 0, i32 3, i32 1029, null, i32 0, i32 45, i32 -1, i32 0, i32 -1}
!47 = !{i32 2}
!48 = !{i32 3}
#version 310 es
// LunarGOO output
precision mediump float; // this will be almost entirely overridden by individual declarations
uniform bvec3 bv;
uniform mediump int imax;
uniform mediump int imin;
uniform mediump uint umax;
uniform mediump uint umin;
flat in mediump ivec3 iv3a;
flat in mediump ivec3 iv3b;
in mediump float f;
in mediump vec3 x;
in mediump vec3 y;
flat in mediump uint uy;
flat in mediump uvec2 uv2c;
flat in mediump uvec2 uv2y;
flat in mediump uvec2 uv2x;
flat in mediump uvec4 uv4y;
flat in mediump ivec4 iv4a;
flat in mediump ivec4 iv4b;
in mediump vec2 v2a;
in mediump vec2 v2b;
in mediump vec4 v4;
out mediump vec4 result;
const float C_0d0 = 0.0;
const vec4 C_vec4p0d0p = vec4(0.0);
const vec2 C_vec2p2d0p = vec2(2.0);
const float C_0d1 = 0.1;

void main()
{
	ivec3 H_9bk0d5 = mix(iv3a,iv3b,bv);
	mediump vec3 H_r2o9bw1 = vec3(H_9bk0d5);
	vec4 H_o3o0001 = vec4(H_r2o9bw1.x, H_r2o9bw1.y, H_r2o9bw1.z, C_0d0);
	mediump vec4 sum = H_o3o0001 + C_vec4p0d0p;
	mediump vec4 H_99dd3e = C_vec4p0d0p + sum;
	mediump vec4 result1 = C_vec4p0d0p + H_99dd3e;
	result = result1;
	mediump vec2 H_67d0kq1 = gl_PointCoord * C_vec2p2d0p;
	vec4 H_0wse3z = vec4(gl_PointCoord.x, gl_PointCoord.y, H_67d0kq1.x, H_67d0kq1.y);
	highp vec4 H_uyk0mz = H_0wse3z + gl_FragCoord;
	highp vec4 result2 = H_uyk0mz + result1;
	result = result2;
	highp float Ll_FragDepth1 = f + C_0d1;
	gl_FragDepth = Ll_FragDepth1;
	
}

